src/transformStateWithClones.js

```diff
@@ -7,7 +7,56 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopies = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        clearObject(stateCopy);
+
+        break;
+
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+
+        break;
+
+      default:
+        continue;
+    }
+
+    stateCopies.push({ ...stateCopy });
+  }
+
+  return stateCopies;
+}
+
+function clearObject(object) {
+  for (const key in object) {
+    delete object[key];
+  }
+
+  // return object.push(object);
+}
+
+function addProperties(object, array) {
```

Comment:
[NAMING] - The parameter `array` in `addProperties` function is misleading since `extraData` is expected to be an object. Please rename it to reflect that it is an object containing properties to add, something like `properties` would be more appropriate.